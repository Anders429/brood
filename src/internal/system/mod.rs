pub(crate) mod schedule;

mod null;

pub(crate) use null::Null;
