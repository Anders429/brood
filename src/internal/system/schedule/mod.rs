pub(crate) mod raw_task;
pub(crate) mod task;
