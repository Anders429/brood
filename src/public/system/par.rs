use crate::{
    query::{filter::Filter, result, view::ParViews},
    registry::Registry,
    world::World,
};

pub trait ParSystem<'a> {
    type Filter: Filter;
    type Views: ParViews<'a> + super::schedule::Claim;

    fn run<R>(&mut self, query_results: result::ParIter<'a, R, Self::Filter, Self::Views>)
    where
        R: Registry + 'a;

    #[inline]
    #[allow(unused_variables)]
    fn world_post_processing<R>(&mut self, world: &mut World<R>)
    where
        R: Registry,
    {
    }
}
