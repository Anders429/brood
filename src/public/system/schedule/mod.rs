pub mod stage;
pub mod task;

mod builder;

pub use builder::Builder;

use crate::{
    internal::system::schedule::{sendable::SendableWorld, raw_task}, registry::Registry,
    world::World,
};
use stage::Stages;

pub struct Schedule<S> {
    stages: S,
}

impl Schedule<stage::Null> {
    pub fn builder() -> Builder<raw_task::Null> {
        Builder::new()
    }
}

impl<'a, S> Schedule<S>
where
    S: Stages<'a>,
{
    pub(crate) fn run<R>(&mut self, world: &'a mut World<R>)
    where
        R: Registry,
    {
        self.stages.run(SendableWorld(world));
    }
}
