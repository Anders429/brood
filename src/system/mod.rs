#[cfg(feature = "parallel")]
pub mod schedule;

mod null;
#[cfg(feature = "parallel")]
mod par;

pub use null::Null;
#[cfg(feature = "parallel")]
pub use par::ParSystem;
#[cfg(feature = "parallel")]
pub use schedule::Schedule;

use crate::{
    query::{filter::Filter, result, view::Views},
    registry::Registry,
    world::World,
};

pub trait System<'a> {
    type Filter: Filter;
    type Views: Views<'a>;

    fn run<R>(&mut self, query_results: result::Iter<'a, R, Self::Filter, Self::Views>)
    where
        R: Registry + 'a;

    #[inline]
    #[allow(unused_variables)]
    fn world_post_processing<R>(&mut self, world: &mut World<R>)
    where
        R: Registry,
    {
    }
}
