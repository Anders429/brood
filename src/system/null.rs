use crate::{
    query::{
        filter,
        filter::Filter,
        result,
        view,
    },
    registry::{
        ContainsViews,
        Registry,
    },
    system::System,
    world::World,
};
#[cfg(feature = "rayon")]
use crate::{
    registry::ContainsParViews,
    system::ParSystem,
};
use core::hint::unreachable_unchecked;

/// A null system.
///
/// As this is an empty `enum`, it can never be instantiated. Its main use is as a generic argument
/// for [`Schedule`]s for the unused generic parameters of [`Stage`]s.
///
/// [`Schedule`]: crate::system::Schedule
/// [`Stage`]: crate::system::schedule::stage::Stage
pub enum Null {}

impl<'a> System<'a> for Null {
    type Filter = filter::None;
    type Views = view::Null;

    fn run<R, FI, VI, P, I, Q>(
        &mut self,
        _query_results: result::Iter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
    ) where
        R: ContainsViews<'a, Self::Views, P, I, Q> + 'a,
        Self::Filter: Filter<R, FI>,
        Self::Views: Filter<R, VI>,
    {
        // SAFETY: This type can never be instantiated. Therefore, this method can never be called.
        unsafe { unreachable_unchecked() }
    }

    fn world_post_processing<R>(&mut self, _world: &mut World<R>)
    where
        R: Registry,
    {
        // SAFETY: This type can never be instantiated. Therefore, this method can never be called.
        unsafe { unreachable_unchecked() }
    }
}

#[cfg(feature = "rayon")]
impl<'a> ParSystem<'a> for Null {
    type Filter = filter::None;
    type Views = view::Null;

    fn run<R, FI, VI, P, I, Q>(
        &mut self,
        _query_results: result::ParIter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
    ) where
        R: ContainsParViews<'a, Self::Views, P, I, Q> + 'a,
        Self::Filter: Filter<R, FI>,
        Self::Views: Filter<R, VI>,
    {
        // SAFETY: This type can never be instantiated. Therefore, this method can never be called.
        unsafe { unreachable_unchecked() }
    }

    fn world_post_processing<R>(&mut self, _world: &mut World<R>)
    where
        R: Registry,
    {
        // SAFETY: This type can never be instantiated. Therefore, this method can never be called.
        unsafe { unreachable_unchecked() }
    }
}
