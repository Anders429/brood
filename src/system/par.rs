use crate::{
    query::{
        filter::Filter,
        result,
        view::ParViews,
    },
    registry::{
        ContainsParQuery,
        Registry,
    },
    world::World,
};

/// An executable type which operates over the entities within a [`World`] in parallel.
///
/// This trait is very similar to the [`System`] trait. The main difference is that the [`run`]
/// method takes a [`result::ParIter`] instead of a [`result::Iter`]. Note that the `Views`
/// associated type must also implement [`ParViews`].
///
/// # Example
/// ``` rust
/// use brood::{
///     query::{
///         filter,
///         filter::Filter,
///         result,
///         views,
///     },
///     registry::ContainsParQuery,
///     system::ParSystem,
/// };
/// use rayon::iter::ParallelIterator;
///
/// // Define components.
/// struct Foo(usize);
/// struct Bar(bool);
///
/// // Define parallel system to operate on those components.
/// struct MySystem;
///
/// impl<'a> ParSystem<'a> for MySystem {
///     type Views = views!(&'a mut Foo, &'a Bar);
///     type Filter = filter::None;
///
///     fn run<R, FI, VI, P, I, Q>(
///         &mut self,
///         query_results: result::ParIter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
///     ) where
///         R: ContainsParQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
///     {
///         query_results.for_each(|result!(foo, bar)| {
///             if bar.0 {
///                 foo.0 += 1;
///             }
///         });
///     }
/// }
/// ```
///
/// [`ParViews`]: crate::query::view::ParViews
/// [`result::Iter`]: crate::query::result::Iter
/// [`result::ParIter`]: crate::query::result::ParIter
/// [`run`]: crate::system::ParSystem::run()
/// [`System`]: crate::system::System
/// [`World`]: crate::world::World
#[cfg_attr(doc_cfg, doc(cfg(feature = "rayon")))]
pub trait ParSystem<'a> {
    /// The filter to apply to queries run by this system.
    type Filter: Filter;
    /// The views on components this system should operate on.
    type Views: ParViews<'a> + Filter;

    /// Logic to be run over the parallel query result.
    ///
    /// Any action performed using the query result should be performed here. If any modifications
    /// to the [`World`] itself are desired based on the query result, those should be performed in
    /// the [`world_post_processing`] method.
    ///
    /// # Example
    /// ``` rust
    /// use brood::{
    ///     query::{
    ///         filter,
    ///         filter::Filter,
    ///         result,
    ///         views,
    ///     },
    ///     registry::ContainsParQuery,
    ///     system::ParSystem,
    /// };
    /// use rayon::iter::ParallelIterator;
    ///
    /// // Define components.
    /// struct Foo(usize);
    /// struct Bar(bool);
    ///
    /// // Define parallel system to operate on those components.
    /// struct MySystem;
    ///
    /// impl<'a> ParSystem<'a> for MySystem {
    ///     type Views = views!(&'a mut Foo, &'a Bar);
    ///     type Filter = filter::None;
    ///
    ///     fn run<R, FI, VI, P, I, Q>(
    ///         &mut self,
    ///         query_results: result::ParIter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
    ///     ) where
    ///         R: ContainsParQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
    ///     {
    ///         query_results.for_each(|result!(foo, bar)| {
    ///             if bar.0 {
    ///                 foo.0 += 1;
    ///             }
    ///         });
    ///     }
    /// }
    /// ```
    ///
    /// [`World`]: crate::world::World
    /// [`world_post_processing`]: crate::system::System::world_post_processing()
    fn run<R, FI, VI, P, I, Q>(
        &mut self,
        query_results: result::ParIter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>,
    ) where
        R: ContainsParQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a;

    /// Logic to be run after processing.
    ///
    /// This is an optional method that can be defined if any changes are desired to be made to the
    /// [`World`] after querying. Changes can be stored using fields of the type implementing
    /// `ParSystem` during the [`run`] method so that they can be accessed by this method.
    ///
    /// # Example
    /// The following example creates a list of entities to remove during evaluation, and then
    /// executes the removal during post processing.
    ///
    /// ``` rust
    /// use brood::{entity, query::{filter, filter::Filter, result, views}, registry::{ContainsParQuery, Registry}, system::ParSystem, World};
    /// use rayon::iter::ParallelIterator;
    ///
    /// // Define components.
    /// struct Foo(usize);
    /// struct Bar(bool);
    ///
    /// // Define system to operate on those components.
    /// struct MySystem {
    ///     // A list of entity identifiers to remove during post processing.
    ///     entities_to_remove: Vec<entity::Identifier>,     
    /// }
    ///
    /// impl<'a> ParSystem<'a> for MySystem {
    ///     type Views = views!(&'a mut Foo, &'a Bar, entity::Identifier);
    ///     type Filter = filter::None;
    ///
    ///     fn run<R, FI, VI, P, I, Q>(&mut self, query_results: result::ParIter<'a, R, Self::Filter, FI, Self::Views, VI, P, I, Q>)
    ///     where
    ///         R: ContainsParQuery<'a, Self::Filter, FI, Self::Views, VI, P, I, Q> + 'a,
    ///     {
    ///         self.entities_to_remove = query_results.filter_map(|result!(foo, bar, entity_identifier)| {
    ///             // If `bar` is true, increment `foo`. Otherwise, remove the entity in post processing.
    ///             if bar.0 {
    ///                 foo.0 += 1;
    ///                 None
    ///             } else {
    ///                 Some(entity_identifier)
    ///             }
    ///         }).collect::<Vec<_>>();
    ///     }
    ///
    ///     fn world_post_processing<R>(&mut self, world: &mut World<R>) where R: Registry {
    ///         for entity_identifier in &self.entities_to_remove {
    ///             world.remove(*entity_identifier);
    ///         }
    ///     }
    /// }
    /// ```
    ///
    /// [`run`]: crate::system::ParSystem::run()
    /// [`World`]: crate::world::World
    #[inline]
    #[allow(unused_variables)]
    fn world_post_processing<R>(&mut self, world: &mut World<R>)
    where
        R: Registry,
    {
    }
}
