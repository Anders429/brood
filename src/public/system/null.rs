#[cfg(feature = "parallel")]
use crate::system::ParSystem;
use crate::{
    query::{filter, result, view},
    registry::Registry,
    system::System,
    world::World,
};
use core::hint::unreachable_unchecked;

pub enum Null {}

impl<'a> System<'a> for Null {
    type Filter = filter::None;
    type Views = view::Null;

    fn run<R>(&mut self, _query_results: result::Iter<'a, R, Self::Filter, Self::Views>)
    where
        R: Registry + 'a,
    {
        unsafe { unreachable_unchecked() }
    }

    fn world_post_processing<R>(&mut self, _world: &mut World<R>)
    where
        R: Registry,
    {
        unsafe { unreachable_unchecked() }
    }
}

#[cfg(feature = "parallel")]
impl<'a> ParSystem<'a> for Null {
    type Filter = filter::None;
    type Views = view::Null;

    fn run<R>(&mut self, _query_results: result::ParIter<'a, R, Self::Filter, Self::Views>)
    where
        R: Registry + 'a,
    {
        unsafe { unreachable_unchecked() }
    }

    fn world_post_processing<R>(&mut self, _world: &mut World<R>)
    where
        R: Registry,
    {
        unsafe { unreachable_unchecked() }
    }
}
