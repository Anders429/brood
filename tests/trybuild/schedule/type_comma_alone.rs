use brood::system::Schedule;

type MySchedule = Schedule!(,);

fn main() {}
