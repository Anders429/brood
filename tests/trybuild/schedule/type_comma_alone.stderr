error: no rules expected the token `,`
 --> tests/trybuild/schedule/type_comma_alone.rs:3:29
  |
3 | type MySchedule = Schedule!(,);
  |                             ^ no rules expected this token in macro call
  |
note: while trying to match meta-variable `$task:ty`
 --> src/system/schedule/mod.rs
  |
  |             ($task:ty $(,$tasks:ty)* $(,)?) => (
  |              ^^^^^^^^
